File: Week 2/4. Count number of Sudoku solutions/sudoku_count.py
N = 9

def findEmptyCell(grid):
    for i in range(N):
        for j in range(N):
            if grid[i][j] == 0:
                return (i, j)
                
    return None
    
def is_valid(grid, row, col, num):
    for i in range(N):
        if grid[row][i] == num:
            return False
            
        if grid[i][col] == num:
            return False
            
        startRow, startCol = row - row % 3, col - col % 3
        for i in range(3):
            for j in range(3):
                if (grid[i + startRow][j + startCol] == num):
                    return False
                    
    return True
    
def count_sudoku_solution(grid):
    empty_cell = findEmptyCell(grid)
    
    if not empty_cell:
        return 1
        
    row, col = empty_cell
    count = 0
    
    for num in range(1, 10):
        if is_valid(grid, row, col, num):
            grid[row][col] = num
            count += count_sudoku_solution(grid)
            grid[row][col] = 0 
    
    return count 
    
board = []
for _ in range(N):
    row = list(map(int, input(). split()))
    board.append(row)
    
print(count_sudoku_solution(board))
    
                    

File: Week 2/4. Count number of Sudoku solutions/sudoku_count.cpp
#include <iostream>

#define N 9
using namespace std;

bool hasEmptyCell(int grid[N][N], int &row, int &col) {  // Pass row and col by reference
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            if (grid[i][j] == 0) {
                row = i;  // Update row
                col = j;  // Update col
                return true;
            }
        }
    }
    return false;
}

bool isValid(int grid[N][N], int row, int col, int num) {
    // Check the row
    for (int i = 0; i < N; ++i) {
        if (grid[row][i] == num) 
            return false;
    }
    
    // Check the col
    for (int i = 0; i < N; ++i) {
        if (grid[i][col] == num) 
            return false;
    }
    
    // Check the 3x3 subgrid
    int startRow = row - row % 3,
        startCol = col - col % 3;
        
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            if (grid[i + startRow][j + startCol] == num)
                return false;
        }
    }
    
    return true;
}

int countSudokuSolutions (int grid[N][N]) {
    int row, col;
    
    if (!hasEmptyCell(grid, row, col)) {
        return 1;  // Solution found
    }
    
    int count = 0;
    for (int num = 1; num <= 9; ++num) {
        if (isValid(grid, row, col, num)) {
            grid[row][col] = num;
            count += countSudokuSolutions(grid);
            grid[row][col] = 0; // Backtracking
        }
    }
    
    return count;
}

int main() {
    int board[N][N];
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            int num;
            cin >> num;
            board[i][j] = num;
        }
    }
    
    cout << countSudokuSolutions(board);
}
